### PR TITLE
Adds LDAP group support

### LINKED ISSUE #4002: LDAP: support team membership management
Our SAML SSO options provide support for managing team membership and roles via SAML groups.

This was not included in our first iteration of LDAP SSO support, but we should fill that gap.

```[tasklist]
### Tasks
- [ ] Research LDAP group membership and how it would apply to FF teams
- [ ] Add team membership management options to LDAP admin dialog
- [ ] Update LDAP auth flow to include team membership management
```

 - https://app-eu1.hubspot.com/contacts/26586079/record/0-2/8175387106
 - https://app-eu1.hubspot.com/contacts/26586079/record/0-2/9822854636

### PR DESCRIPTION
**Pull request by hardillb**
fixes #4002 
## Description

Adds support for controlling team membership with LDAP groups

I've tested this with my local LDAP instance.

Updated UI

image

It reuses parts from the SAML group support where possible.

There may be some code duplication in the group managemnts, but as they exist in different areas I didn't re-factor them out to a shared function.

## Related Issue(s)

#4002

## Checklist

 - [x] I have read the [contribution guidelines](https://github.com/FlowFuse/flowfuse/blob/main/CONTRIBUTING.md)
 - [ ] Suitable unit/system level tests have been added and they pass 
 - [x] Documentation has been updated
    - [ ] Upgrade instructions
    - [x] Configuration details
    - [ ] Concepts
 - [ ] Changes `flowforge.yml`?
    - [ ] Issue/PR raised on `FlowFuse/helm` to update ConfigMap Template
    - [ ] Issue/PR raised on `FlowFuse/CloudProject` to update values for Staging/Production

## Labels

 - [ ] Includes a DB migration? -> add the `area:migration` label

### CODE REVIEW

File: docs/admin/sso/ldap.md [hunk 1]
```diff
@@ -69,4 +69,62 @@ option in the SOO configuration.
...
 to create the username. The user will be directed to their settings page where they
-can modify their user details to their preferred values.
\ No newline at end of file
+can modify their user details to their preferred values.
+
+## Managing Team Membership with LDAP Groups
+
+LDAP implementations can also be used to group users
+
+To enable tis option, select the `Manage roles using group assertions` in the SSO configuration.
```
Comment (Steve-Mcl) [inline docs/admin/sso/ldap.md:78]:
```suggestion
To enable this option, select the `Manage roles using group assertions` in the SSO configuration.
```

File: forge/ee/lib/sso/index.js [hunk 2]
```diff
@@ -424,6 +428,130 @@ module.exports.init = async function (app) {
         }
     }
 
+    // LDAP Group Membership
+    async function updateTeamMembershipLDAP (adminClient, user, userDN, providerOpts) {
+        const filter = `(|(uniqueMember=${userDN})(member=${userDN}))`
+        // const { searchEntries } = await adminClient.search(providerOpts.options.groupsDN, {
+        const { searchEntries } = await adminClient.search('ou=groups,dc=hardill,dc=me,dc=uk', {
```
Comment (Steve-Mcl) [inline forge/ee/lib/sso/index.js:435]: temp testing?

File: docs/admin/sso/ldap.md [hunk 3]
```diff
@@ -69,4 +69,62 @@ option in the SOO configuration.
...
+ - `dashboard`
+
+*Note*: this uses the team slug property to identify the team. This has been chosen to simplify managing
+the groups in the LDAP Provider - rather than using the team's id. However, a team's slug can be changed
+by a team owner. Doing so will break the link between the group and the team membership - so should only
+be done with care.
+
+## Managing Admin users
+
+The SSO Configuration can be configured to managed the admin users of the platform by enabling the
```
Comment (Steve-Mcl) [inline docs/admin/sso/ldap.md:117]:
```suggestion
The SSO Configuration can be configured to manage the admin users of the platform by enabling the
```

File: frontend/src/pages/admin/Settings/SSO/createEditProvider.vue [hunk 4]
```diff
@@ -93,7 +93,63 @@
...
+                            <FormRow v-if="input.options.groupAllTeams === false" v-model="input.options.groupOtherTeams" type="checkbox" class="pl-4">
+                                Allow users to be in other teams
+                                <template #description>
+                                    If enabled, users can be members of any teams not listed above and their membership/roles are not managed
+                                    by this SSO configuration.
+                                </template>
+                            </FormRow>
+                            <FormRow v-model="input.options.groupAdmin" type="checkbox">Manage Admin roles using group assertions</FormRow>
+                            <FormRow v-if="input.options.groupAdmin" v-model="input.options.groupAdminName" :error="groupAdminNameError" class="pl-4">Admin Users SAML Group name</FormRow>
+                        </div> -->
```
Comment (Steve-Mcl) [inline frontend/src/pages/admin/Settings/SSO/createEditProvider.vue:119]:
```suggestion
```

Is all this needed for some future thing?

**Review by Steve-Mcl — CHANGES_REQUESTED** [review]

File: frontend/src/pages/admin/Settings/SSO/createEditProvider.vue (hunk 4 above)
Comment (hardillb) [reply frontend/src/pages/admin/Settings/SSO/createEditProvider.vue:119]: No, just forgot to remove old version

Comment (hardillb) [inline frontend/src/pages/admin/Settings/SSO/createEditProvider.vue:119]:
```suggestion
```

**Review by Steve-Mcl — COMMENTED** [review]
Code wise, it all looks good however I have no way of trying this out and in the spirit of "done" I have to ask about tests. There is [no coverage](https://app.codecov.io/gh/FlowFuse/flowfuse/pull/4407/blob/forge/ee/lib/sso/index.js#L57) for the new functionality

Happy to approve if there is a follow up raised and @joepavitt agrees.

**Comment by hardillb** [comment]
@Steve-Mcl all the new functions require a live LDAP server.

And Nick is probably the only other person with a LDAP setup to test against.

**Comment by Steve-Mcl** [comment]
> @Steve-Mcl all the new functions require a live LDAP server.
> 
> And Nick is probably the only other person with a LDAP setup to test against.

Indeed which is why I would be ok with approving this so long as there is a follow up task and @joepavitt is on-board.

**Comment by joepavitt** [comment]
Agreed. Happy to proceed

**Review by Steve-Mcl — APPROVED** [review]
Approved pending follow up issue raised to bring test coverage up

File: forge/ee/lib/sso/index.js [hunk 5]
```diff
@@ -424,6 +428,129 @@ module.exports.init = async function (app) {
         }
     }
 
+    // LDAP Group Membership
+    async function updateTeamMembershipLDAP (adminClient, user, userDN, providerOpts) {
+        const filter = `(|(uniqueMember=${userDN})(member=${userDN}))`
+        const { searchEntries } = await adminClient.search(providerOpts.options.groupsDN, {
```
Comment (hardillb) [inline forge/ee/lib/sso/index.js:434]:
```suggestion
        const { searchEntries } = await adminClient.search(providerOpts.groupsDN, {
```

**Comment by hardillb** [comment]
@Steve-Mcl  https://github.com/FlowFuse/flowfuse/issues/4416

**Pull request merged**